src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala

```diff
@@ -45,24 +45,45 @@ import scala.collection.mutable.ListBuffer
 
 
 object GroupReadsByUmi {
+  /** A type representing an actual UMI that is a string of DNA sequence. */
+  type Umi = String
+  /** A type to represent a unique ID assigned to a molecule. */
+  type MoleculeId = String
+
   private val SortPositionAttribute        = "__GRs"
   private val LibraryAttribute             = "__GRlb"
   private val InsertEndsAndStrandAttribute = "__Insert"
 
   /** Trait that can be implemented to provide a UMI assignment strategy. */
   private[umi] sealed trait UmiAssigner {
-    def assign(rawUmis: Seq[String]) : Map[String, String]
+    private val counter = new AtomicLong()
+
+    /** Take in a sequence of UMIs and assign each UMI to a unique UMI group ID. */
+    def assign(rawUmis: Seq[Umi]) : Map[Umi, MoleculeId]
```

Comment:
What's interesting is that it looks like you want to force folks to use `assignIds`, but it can be omitted by an implementing class. Perhaps you want to have:
```
def assign(rawUmis: Seq[Umi]) : Map[Umi, MoleculeId] = {
  assignIds(umiToSentinelUmi(rawUmis))
}

protected def umiToSentinelUmi(rawUmis: Seq[Umi]): Map[Umi,Umi]
```
?